samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -18,47 +18,178 @@
+        if (container != null && container.getStatus().equals(SamzaContainerStatus.RUNNING)) {
+          container.shutdown(true);
+          boolean shutdownComplete = false;
+          try {
+            shutdownComplete = jcContainerLatch.await(taskShutdownMs, TimeUnit.MILLISECONDS);
+          } catch (InterruptedException e) {
+            LOGGER.warn("Container shutdown was interrupted!" + container.toString());
+          }
+          if (!shutdownComplete) {
+            LOGGER.warn("Container " + container.toString() + " may not have shutdown successfully. Stopping the processor.");
+            container = null;
+            stop();
+          } else {
+            LOGGER.info("Container " + container.toString() + " shutdown successfully");
+          }
+        }
+      }
+
+      @Override
+      public void onNewJobModel(String processorId, JobModel jobModel) {
+        if (!jobModel.getContainers().containsKey(processorId)) {
+          LOGGER.warn("JobModel did not contain the processorId. Stopping the processor.");
+          stop();
+        } else {
+          jcContainerLatch = new CountDownLatch(1);
+
+          SamzaContainerListener containerListener = new SamzaContainerListener() {
+            @Override
+            public void onContainerStart() {
+              if (!processorOnStartCalled) {
+                // processorListener is called on start only the first time the container starts.
+                // It is not called after every re-balance of partitions among the processors
+                processorOnStartCalled = true;
+                if (processorListener != null) {
+                  processorListener.onStart();
+                }
+              }
+            }
+
+            @Override
+            public void onContainerStop(boolean pauseByJm) {
+              if (jcContainerLatch != null) {
+                jcContainerLatch.countDown();
+              } else {
+                LOGGER.warn("JobCoordinatorLatch was null. It is possible for some component to be waiting.");
+              }
+
+              if (pauseByJm) {
+                LOGGER.info("Container " + container.toString() + " stopped due to a request from JobCoordinator.");
+              } else {  // sp.stop was called or container stopped by itself
+                LOGGER.info("Container " + container.toString() + " stopped.");
+                container = null; // this guarantees that stop() doesn't try to stop container again
+                stop();
+              }
+            }
+
+            @Override
+            public void onContainerFailed(Throwable t) {
+              if (jcContainerLatch != null) {
+                jcContainerLatch.countDown();
+              } else {
+                LOGGER.warn("JobCoordinatorLatch was null. It is possible for some component to be waiting.");
+              }
+              LOGGER.info("Container failed with " + t + ". Stopping the processor.");
+              stop();
```

Comment:
Yeah. I already added that. Missed it in the first round. Thanks for pointing it out!